src/main/java/pl/futurecollars/invoice/db/sql/SqlDatabase.java

```diff
@@ -0,0 +1,327 @@
+  }
+
+  @Override
+  @Transactional
+  public long save(Invoice invoice) {
+    GeneratedKeyHolder keyHolder = new GeneratedKeyHolder();
+
+    long buyerId = insertCompany(invoice.getBuyer());
+    long sellerId = insertCompany(invoice.getSeller());
+    long invoiceId = insertInvoice(invoice, buyerId, sellerId);
+
+    invoice.getInvoiceEntries().forEach(entry -> {
+      jdbcTemplate.update(connection -> {
+        PreparedStatement ps = connection
+            .prepareStatement(
+                "insert into invoice_entries (description, quantity, net_price, vat_value, vat_rate, expense_related_to_car) "
+                    + "values (?, ?, ?, ?, ?, ?);",
+                new String[] {"id"});
+        ps.setString(1, entry.getDescription());
+        ps.setBigDecimal(2, entry.getQuantity());
+        ps.setBigDecimal(3, entry.getNetPrice());
+        ps.setBigDecimal(4, entry.getVatValue());
+        ps.setInt(5, vatToId.get(entry.getVatRate()));
+        ps.setObject(6, insertCarAndGetItId(entry.getExpenseRelatedToCar()));
+        return ps;
+      }, keyHolder);
+
+      long invoiceEntryId = Objects.requireNonNull(keyHolder.getKey()).longValue();
+
+      jdbcTemplate.update(connection -> {
+        PreparedStatement ps = connection.prepareStatement(
+            "insert into invoice_invoice_entry (invoice_id, invoice_entry_id) values (?, ?);");
+        ps.setLong(1, invoiceId);
+        ps.setLong(2, invoiceEntryId);
+        return ps;
+      });
+    });
+    return invoiceId;
+  }
+
+  private int insertInvoice(Invoice invoice, long buyerId, long sellerId) {
+    GeneratedKeyHolder keyHolder = new GeneratedKeyHolder();
+    jdbcTemplate.update(connection -> {
+      PreparedStatement ps =
+          connection.prepareStatement("insert into invoices (date, number, buyer, seller) values (?, ?, ?, ?);", new String[] {"id"});
+      ps.setDate(1, Date.valueOf(invoice.getDate()));
+      ps.setString(2, invoice.getNumber());
+      ps.setLong(3, buyerId);
+      ps.setLong(4, sellerId);
+      return ps;
+    }, keyHolder);
+    return Objects.requireNonNull(keyHolder.getKey()).intValue();
+  }
+
+  private long insertCompany(Company buyer) {
+    GeneratedKeyHolder keyHolder = new GeneratedKeyHolder();
+
+    jdbcTemplate.update(connection -> {
+      PreparedStatement ps = connection.prepareStatement(
+          "insert into companies (name, address, tax_identification_number, health_insurance, pension_insurance) values (?, ?, ?, ?, ?);",
+          new String[] {"id"});
+      ps.setString(1, buyer.getName());
+      ps.setString(2, buyer.getAddress());
+      ps.setString(3, buyer.getTaxIdentificationNumber());
+      ps.setBigDecimal(4, buyer.getHealthInsurance());
+      ps.setBigDecimal(5, buyer.getPensionInsurance());
+      return ps;
+    }, keyHolder);
+    return Objects.requireNonNull(keyHolder.getKey()).longValue();
+  }
+
+  private Integer insertCarAndGetItId(Car car) {
+    if (car == null) {
+      return null;
```

Comment:
Lepiej rzucić wyjątkiem w tej sytuacji.

Reply:
Gdy car == null to faktura nie dotyczy samochodu, więc sytuacja jest prawidłowa.